### ISSUE TITLE
Longevity test release-1.6 CN OOM

### ISSUE THREAD

**Issue by huangjw806** [risingwavelabs/risingwave#14712]
```
"!!! longevity Result!!!"
Today's date:2024-01-22
Result               FAIL                
Pipeline Message     Test v1.6.1-rc      
TestBed              kubebench/3264g-medium-3cn-all-affinity
RW Version           v1.6.1-rc           
Test Start time      2024-01-21 17:00:08 
Test End time        2024-01-22 05:02:24 
Namespace            usrlngvty-20240121-165055
Queries              nexmark_q0,nexmark_q1,nexmark_q2,nexmark_q3,nexmark_q4,nexmark_q5,nexmark_q7,nexmark_q8,nexmark_q9,nexmark_q10,nexmark_q12,nexmark_q14,nexmark_q15,nexmark_q16,nexmark_q17,nexmark_q18,nexmark_q19,nexmark_q20,nexmark_q21,nexmark_q22,nexmark_q101,nexmark_q102,nexmark_q103,nexmark_q104,nexmark_q105
Grafana Metric       https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&var-namespace=usrlngvty-20240121-165055&from=1705856408000&to=1705899744000
Grafana Logs         https://grafana.test.risingwave-cloud.xyz/d/liz0yRCZz1/log-search-dashboard?orgId=1&var-data_source=Logging:%20test-useast1-eks-a&var-namespace=usrlngvty-20240121-165055&from=1705856408000&to=1705899744000
Memory Dumps to S3   https://s3.console.aws.amazon.com/s3/buckets/test-useast1-mgmt-bucket-archiver?region=us-east-1&prefix=k8s/usrlngvty-20240121-165055/&showversions=false
Buildkite Job        https://buildkite.com/risingwave-test/longevity-test/builds/954
Crash Container logs https://rw-qa-artifacts-public.s3.us-east-1.amazonaws.com/longevity/954_logs.txt
Report               https://rw-qa-artifacts-public.s3.us-east-1.amazonaws.com/longevity/954_report.txt


================================================================================
Restarted/Crashed Containers Details 
================================================================================
CONTAINER crashed/Restarted: benchmark-risingwave-compute-c-0 restart_count:1  phase:Running status:True
CONTAINER crashed/Restarted: benchmark-risingwave-compute-c-2 restart_count:1  phase:Running status:True
```

**Comment by BugenZhao** [comment]
Collapsed files: https://buildkite.com/risingwave-test/generate-collapsed-heap-files/builds/14#018d2fc4-1df1-435d-85b5-58d88d1a990d

**Comment by huangjw806** [comment]
rerun include `nexmark_q6_group_top1`

```
"!!! longevity Result!!!"
Today's date:2024-01-22
Result               FAIL                
Pipeline Message     longevity test v1.6.1-rc
TestBed              kubebench/3264g-medium-3cn-all-affinity
RW Version           v1.6.1-rc           
Test Start time      2024-01-22 05:39:43 
Test End time        2024-01-22 15:01:42 
Namespace            usrlngvty-20240122-053033
Queries              nexmark_q0,nexmark_q1,nexmark_q2,nexmark_q3,nexmark_q4,nexmark_q5,nexmark_q6_group_top1,nexmark_q7,nexmark_q8,nexmark_q9,nexmark_q10,nexmark_q12,nexmark_q14,nexmark_q15,nexmark_q16,nexmark_q17,nexmark_q18,nexmark_q19,nexmark_q20,nexmark_q21,nexmark_q22,nexmark_q101,nexmark_q102,nexmark_q103,nexmark_q104,nexmark_q105
Grafana Metric       https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&var-namespace=usrlngvty-20240122-053033&from=1705901983000&to=1705935702000
Grafana Logs         https://grafana.test.risingwave-cloud.xyz/d/liz0yRCZz1/log-search-dashboard?orgId=1&var-data_source=Logging:%20test-useast1-eks-a&var-namespace=usrlngvty-20240122-053033&from=1705901983000&to=1705935702000
Memory Dumps to S3   https://s3.console.aws.amazon.com/s3/buckets/test-useast1-mgmt-bucket-archiver?region=us-east-1&prefix=k8s/usrlngvty-20240122-053033/&showversions=false
Buildkite Job        https://buildkite.com/risingwave-test/longevity-test/builds/956
Crash Container logs https://rw-qa-artifacts-public.s3.us-east-1.amazonaws.com/longevity/956_logs.txt
Report               https://rw-qa-artifacts-public.s3.us-east-1.amazonaws.com/longevity/956_report.txt


================================================================================
Restarted/Crashed Containers Details 
================================================================================
CONTAINER crashed/Restarted: benchmark-risingwave-compute-c-0 restart_count:25  phase:Running status:True
```

**Comment by lmatz** [comment]
Collapsed files: https://buildkite.com/risingwave-test/generate-collapsed-heap-files/builds/17

**Comment by lmatz** [comment]
>Collapsed files: https://buildkite.com/risingwave-test/generate-collapsed-heap-files/builds/17

![SCR-20240123-ksv](https://github.com/risingwavelabs/risingwave/assets/5791930/5e8d5052-f4ca-48be-a17f-4961e63c7770)


1. MZ executor uses 1.4GB, does it make sense?
2. And the total memory only uses less than 10GB, but somehow it generates a dump, does it make sense?

**Comment by lmatz** [comment]
![SCR-20240123-lbm](https://github.com/risingwavelabs/risingwave/assets/5791930/cd9d1648-0f53-4125-8cc9-3f2dff19b66a)

is it because if we set `request` to 13GB also, the pod cannot be scheduled? cc: @huangjw806

**Comment by fuyufjh** [comment]
> And the total memory only uses less than 10GB, but somehow it generates a dump, does it make sense?

Auto memory dump is triggered by OS memory usage (rss), while the heap dump is made by jemalloc, which is usually less than the rss for up to 30%.

**Comment by lmatz** [comment]
is it normal that only the materialize executor takes a large chunk of memory? cannot find other executors using comparable amount of memory
I suppose the materialize executor does not need any cache in it, only temporary memory allocation which disappears after the chunk gets flushed into the storage?

**Comment by fuyufjh** [comment]
> is it normal that only the materialize executor takes a large chunk of memory? cannot find other executors using comparable amount of memory

Ohh, indeed. The `Materialize` executor should not validate data here. Need to dive in.

**Comment by huangjw806** [comment]
> is it because if we set `request` to 13GB also, the pod cannot be scheduled? 
I think it’s work, because it seems that the total resources have not reached the limit.

> Whether a pod runs into OOM depends on the `limit`, right
yes.

**Comment by huangjw806** [comment]
The total memory of the machine also seems to be sufficient.
<img width="952" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/29391315/ea5844ee-9fb6-40e7-8a92-baf61878a679">

node exporter: https://grafana.test.risingwave-cloud.xyz/d/LDL_mx9nz123/kubernetes-node-exporter?orgId=1&var-DS_PROMETHEUS=P2453400D1763B4D9&var-node=ip-10-0-43-102.ec2.internal&var-internal_ip=10.0.43.102&var-diskdevices=%5Ba-z%5D%2B%7Cnvme%5B0-9%5D%2Bn%5B0-9%5D%2B%7Cmmcblk%5B0-9%5D%2B&from=1705893102767&to=1705983763575

**Comment by arkbriar** [comment]
The process was killed by the OOM killer. You can find the logs [here](https://grafana.test.risingwave-cloud.xyz/explore?panes=%7B%22m9v%22:%7B%22datasource%22:%22PE59595AED52CF917%22,%22queries%22:%5B%7B%22refId%22:%22A%22,%22expr%22:%22%7Btype%3D%5C%22journal%5C%22,%20hostname%3D%5C%22ip-10-0-43-102.ec2.internal%5C%22,%20transport%3D%5C%22kernel%5C%22%7D%20%7C%3D%20%60%60%20%7C%20json%20%7C%20line_format%20%60%7B%7B.MESSAGE%7D%7D%60%22,%22queryType%22:%22range%22,%22datasource%22:%7B%22type%22:%22loki%22,%22uid%22:%22PE59595AED52CF917%22%7D,%22editorMode%22:%22builder%22%7D%5D,%22range%22:%7B%22from%22:%221705920001804%22,%22to%22:%221705976608701%22%7D%7D%7D&schemaVersion=1&orgId=1).

```plain
oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=cri-containerd-89ddc6d34df8009759c8a68262451ea93278c07d6330a80b7c8504bc57be1196.scope,mems_allowed=0,oom_memcg=/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod455a7760_dc45_4b38_b624_977c170dd1d9.slice,task_memcg=/kubepods.slice/kubepods-burstable.slice/kubepods-burstable-pod455a7760_dc45_4b38_b624_977c170dd1d9.slice/cri-containerd-89ddc6d34df8009759c8a68262451ea93278c07d6330a80b7c8504bc57be1196.scope,task=risingwave,pid=733811,uid=0

Memory cgroup out of memory: Killed process 733811 (risingwave) total-vm:44901588kB, anon-rss:13531468kB, file-rss:396436kB, shmem-rss:0kB, UID:0 pgtables:77204kB oom_score_adj:838
```

**Comment by fuyufjh** [comment]
> > is it normal that only the materialize executor takes a large chunk of memory? cannot find other executors using comparable amount of memory
> 
> Ohh, indeed. The `Materialize` executor should not validate data here. Need to dive in.

No. We made a mistake. The ~1.5GB memory was not taken by Materialize executor itself, but by the children executors. Rust's async is based on polling, so the children executors are under Materialize in the stack tree.

In the below graph, the highlighted boxes contain keyword "stream::executor::" i.e. they are all stream executors.

<img width="760" alt="image" src="https://github.com/risingwavelabs/risingwave/assets/10192522/d1559771-7d64-45b2-b2e8-ff6c97480ab6">

**Comment by fuyufjh** [comment]
I think this issue is same as https://github.com/risingwavelabs/risingwave/issues/14736. See the analysis at https://github.com/risingwavelabs/risingwave/issues/14736#issuecomment-1905247180.

**Comment by st1page** [comment]
It could because we disable the memtable spill because of the inconsistent issue
on release-1.6 https://github.com/risingwavelabs/risingwave/commit/06058f046de38e91078c1d54f880eccfc934f3de
on main https://github.com/risingwavelabs/risingwave/commit/89a8297ff9082737ccf01f853e2ea3458413de28

**Comment by fuyufjh** [comment]
Rerun https://buildkite.com/risingwave-test/longevity-test/builds/980
with `imm_merge_threshold`=4 and `mem_table_spill_threshold`=4MB, and disabled `check_compaction_result`

**Comment by hzxa21** [comment]
May be due to an extra copy of block during decode after #13558. The fix #14786 just gets merged. Let's wait for today's longevity to see whether the situation is improved.

**Comment by huangjw806** [comment]
`nightly-20240125` cn still oom.

```
================================================================================
longevity-test Result
================================================================================
Result               FAIL                
Pipeline Message     @Nightly run all nexmark (8 sets of nexmark queries) with 10k throughput
Namespace            reglngvty-20240125-150209
TestBed              medium-arm-3cn-all-affinity
RW Version           nightly-20240125    
Test Start time      2024-01-25 15:11:13 
Test End time        2024-01-26 03:13:24 
Test Queries         nexmark_q0,nexmark_q1,nexmark_q2,nexmark_q3,nexmark_q4,nexmark_q5,nexmark_q6_group_top1,nexmark_q7,nexmark_q8,nexmark_q9,nexmark_q10,nexmark_q12,nexmark_q14,nexmark_q15,nexmark_q16,nexmark_q17,nexmark_q18,nexmark_q19,nexmark_q20,nexmark_q21,nexmark_q22,nexmark_q101,nexmark_q102,nexmark_q103,nexmark_q104,nexmark_q105
Grafana Metric       https://grafana.test.risingwave-cloud.xyz/d/EpkBw5W4k/risingwave-dev-dashboard?orgId=1&var-datasource=Prometheus:%20test-useast1-eks-a&var-namespace=reglngvty-20240125-150209&from=1706195473000&to=1706238804000
Grafana Logs         https://grafana.test.risingwave-cloud.xyz/d/liz0yRCZz1/log-search-dashboard?orgId=1&var-data_source=Logging:%20test-useast1-eks-a&var-namespace=reglngvty-20240125-150209&from=1706195473000&to=1706238804000
Memory Dumps         https://s3.console.aws.amazon.com/s3/buckets/test-useast1-mgmt-bucket-archiver?region=us-east-1&bucketType=general&prefix=k8s/reglngvty-20240125-150209/&showversions=false
Buildkite Job        https://buildkite.com/risingwave-test/longevity-test/builds/985

================================================================================
Restarted/Crashed Pods Details
 ================================================================================
Pod crashed/Restarted: benchmark-risingwave-compute-c-2 restart_count:2  phase:Running status:True
```
